### PR TITLE
Switch http client query encoding

### PR DESCRIPTION
**Pull request by RobRikken**
Fix for the issue #15497.

This consists of the addition of the encoding query option and the tests to make sure it works.

### CODE REVIEW

File: src/Http/Client.php [hunk 1]
```diff
@@ -519,20 +520,26 @@ public function buildUrl(string $url, $query = [], array $options = []): string
...
-            $url .= $q;
-            $url .= is_string($query) ? $query : http_build_query($query);
-        }
         $defaults = [
             'host' => null,
             'port' => null,
             'scheme' => 'http',
             'basePath' => '',
             'protocolRelative' => false,
+            'query_encoding' => PHP_QUERY_RFC1738,
```
Comment (markstory) [inline src/Http/Client.php:529]: Could this be `queryEncoding`? Do we need an option at all. Seems like we could/should default to 3986 given that 1738 is updated by 3986.

**Review by markstory — COMMENTED** [review]
Thanks for sending this pull request. I'm not sure we need to have strict compatibility here. Servers should generally accept percent encoding.

File: src/Http/Client.php (hunk 1 above)
Comment (RobRikken) [reply src/Http/Client.php:529]: No problem, I've changed the name to  queryEncoding.

**Comment by RobRikken** [comment]
The default is now changed to PHP_QUERY_RFC3986, this does however give me the following error in the test suite:

```
3) Cake\Test\TestCase\Http\ClientTest::testHeadQuerystring
Expectation failed for method name is "send" when invoked 1 time(s)
Failed asserting that two strings are identical.
--- Expected
+++ Actual
@@ @@
-'http://cakephp.org/search?q=hi+there'
+'http://cakephp.org/search?q=hi%20there'
```

I however do agree that %20 should be generally accepted.

**Comment by ADmad** [comment]
I too don't think the `queryEncoding` option is necessary, we should simply switch to PHP_QUERY_RFC3986 and update existing test cases.

**Comment by RobRikken** [comment]
I have set PHP_QUERY_RFC3986 as the only option and updated the testcases to now use the %20 instead of +.

**Comment by othercorey** [comment]
Because this is in 4.3, we should have an entry in the migration notes.

**Pull request merged**